src/systems/filecoin_blockchain/chainsync/_index.md

```diff
@@ -1,58 +1,388 @@
+            c.Graphsync.Pull(c.Peers, sel, c.IpldStore)
+            // Pull SHOULD NOT pull what c.IpldStore already has (check first)
+            // Pull SHOULD pull from different peers simultaneously
+            // Pull SHOULD be efficient (try different parts of the tree from many peers)
+            // Graphsync implementations may not offer these features. These features
+            // can be implemented on top of a graphsync that only pulls from a single
+            // peer and does not check local store first.
+        }
+        c.ChainCatchup() // on to CHAIN_CATCHUP
+    }
+    ```
+- **security conditions to transition out:**
+  - `StateTree` for `LastTrustedCheckpoint` MUST be stored locally and verified (hashing is enough)
+- **transitions out:**
+  - once node receives and verifies complete `StateTree` for `LastTrustedCheckpoint`: move to `CHAIN_CATCHUP`
+
+## ChainSync FSM: `CHAIN_CATCHUP`
+
+- While in this state:
+  - `ChainSync` is well-bootstrapped, and has an initial **trusted** `StateTree` to start from.
+  - `ChainSync` is receiving latest `Blocks` from `BlockPubsub`
+  - `ChainSync` starts fetching and validating blocks (see _Block Fetching and Validation_ above).
+  - `ChainSync` has unvalidated blocks between `ChainSync.FinalityTipset` and `ChainSync.TargetHeads`
+- **Chain State and Finality**:
+  - In this state, the **chain MUST NOT advance** beyond whatever the node already has.
+  - No new blocks are reported to consumers.
+  - The chain state provided is the available `Blocks` and `StateTree` for all available epochs,
+    specially the `FinalityTipset`.
+- **security conditions to transition out:**
+  - Gaps between `ChainSync.FinalityTipset ... ChainSync.BestTargetHead` have been closed:
+    - All `Blocks` and their content MUST be fetched, stored, linked, and validated locally.
+      This includes `BlockHeaders`, `Messages`, etc.
+    - Bad heads have been expunged from `ChainSync.BestHeads`. Bad heads include heads that initially
+      seemed good but turned out invalid, or heads that `ChainSync` has failed to connect (ie. cannot
+      fetch ancestors connecting back to `ChainSync.FinalityTipset` within a reasonable amount of time).
+    - All blocks between `ChainSync.FinalityTipset ... ChainSync.BestHeads` have been validated
+      This means all blocks _before_ the best heads.
+  - Not under a temporary network partition
+- **transitions out:**
+  - once gaps between `ChainSync.FinalityTipset ... ChainSync.BestHeads` are closed: move to `CHAIN_FOLLOW`
+
+## ChainSync FSM: `CHAIN_FOLLOW`
+
+- While in this state:
+  - `ChainSync` is well-bootstrapped, and has an initial **trusted** `StateTree` to start from.
+  - `ChainSync` fetches and validates blocks (see _Block Fetching and Validation_).
+  - `ChainSync` is receiving and validating latest `Blocks` from `BlockPubsub`
+  - `ChainSync` DOES NOT have unvalidated blocks between `ChainSync.FinalityTipset` and `ChainSync.TargetHeads`
```

Comment:
As soon as a new block is received by gossip, does `TargetHeads` not change? So then that new block would be unvalidated.